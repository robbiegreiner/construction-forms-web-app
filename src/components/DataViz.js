/* eslint-disable no-unused-vars, array-callback-return , default-case*/
import React, { Component } from 'react';
import { Route } from 'react-router-dom';
import { 
  fetchProjects, 
  fetchEmployees, 
  fetchHotwork, 
  fetchPretask } from '../utils/fetchUtils';
import Card from './Card';
import HotworkForm from './HotworkForm';
import PretaskForm from './PretaskForm';

import '../styles/DataViz.css';

class DataViz extends Component {

  constructor() {
    super();
    this.state = {
      projects: [],
      employees: [],
      hotworkForms: [],
      pretaskForms: [],
      dataVizControl: null,
      selectedInfo: {}
    };
  }

  componentDidMount() {
    fetchProjects(this);
    fetchEmployees(this);
    fetchHotwork(this);
    fetchPretask(this);
  }

  reportsBySelectionHandler = (selection) => {
    this.setState({ dataVizControl: selection });
  }

  cardClickHandler = (selectedInfo) => {
    this.setState({ selectedInfo });
  }

  // eslint-disable-next-line no-unused-vars
  mockCardClickHandler = (selectedInfo) => null;

  closeSelected = () => {
    this.setState({ selectedInfo: {} });
  }

  renderSelected = (selectedInfo) => {
    if (this.props.selectedControl === '' && Object.keys(selectedInfo).length !== 0) {
      this.setState({ selectedInfo: {} });
    }
    return (
      <div className="selected-div">
        {
          selectedInfo.employeeInfo &&
             <div>
               <h5>{selectedInfo.header}</h5>
               <p>{selectedInfo.body}</p>
               <p>{selectedInfo.employeeInfo[0]}</p>
               <p>{selectedInfo.employeeInfo[1]}</p>
               <button onClick={this.closeSelected}>X</button>
             </div>
        }
        {
          selectedInfo.projectInfo &&
           <div>
             <h5>{selectedInfo.header}</h5>
             <p>{selectedInfo.body}</p>
             <p>Public: {selectedInfo.projectInfo[0] ? 'Yes' : 'No'}</p>
             <p>Union: {selectedInfo.projectInfo[1] ? 'Yes' : 'No'}</p>
             <button onClick={this.closeSelected}>X</button>
           </div>
        }
        {
          selectedInfo.formDetails &&
             <div>
               <h5>{selectedInfo.header}</h5>
               <p>{selectedInfo.body}</p>
               <p>Employee Email: {selectedInfo.formDetails[0]}</p>
               <p>Project ID: {selectedInfo.formDetails[1]}</p>
               <p>Company: {selectedInfo.formDetails[2]}</p>
               <p>Date: {selectedInfo.formDetails[3]}</p>
               <p>Firewatch Requirement: {selectedInfo.formDetails[4]}</p>
               <p>Area Inspected: {selectedInfo.formDetails[5] ? 'Yes' : 'No'}</p>
               <p>Fire Extinguisher: {selectedInfo.formDetails[6] ? 'Yes' : 'No'}</p>
               <p>Flammables Removed: {selectedInfo.formDetails[7]? 'Yes' : 'No'}</p>
               <p>Sprinkler Heads Protected: {selectedInfo.formDetails[8] ? 'Yes' : 'No'}</p>
               <img src={selectedInfo.formDetails[9]} alt={selectedInfo.header + ' Signature'}/>
               <button onClick={this.closeSelected}>X</button>
             </div>
        }
      </div>
    );
  }

  renderWelcome = (selectedControl) => {
    if (!selectedControl) {
      return (
        <div className="welcome">
          <h1>Welcome to Construction Forms!</h1>
          <div className="logo"></div>
        </div>
      );
    }
  }

  aToZ = arrayOfObjs => {
    const sorted = arrayOfObjs.sort((a, b) => a.name.localeCompare(b.name));
    return sorted;
  }

  cardSwitch = (selectedControl) => {
    switch (selectedControl) {
    case 'Projects':
      return this.aToZ(this.state.projects).map((project, index) => {
        return <Card
          key={'project' + index}
          header={project.name}
          body={project.location}
          projectInfo={[project.public, project.union]}
          cardClickHandler={this.cardClickHandler}
        />;
      });
    case 'Employees':
      return this.aToZ(this.state.employees).map((employee, index) => {
        return <Card
          key={'employee' + index}
          header={employee.name}
          body={employee.position}
          employeeInfo={[employee.phone, employee.email]}
          cardClickHandler={this.cardClickHandler}
        />;
      });
    case 'All Reports':
      return this.state.hotworkForms.concat(this.state.pretaskForms).map((form, index) => {
        const sigImg = new Image();
        sigImg.src = form.signature;
        return <Card
          key={'form' + index}
          header={
            Object.keys(form).includes('firewatchRequirement') ? 'Hotwork Permit - ' + form.employee_name : 'Pretask Permit - ' + form.employee_name
          }
          cardClickHandler={this.cardClickHandler}
          formDetails={[
            form.employee_email,
            form.project_id,
            form.company,
            form.date,
            form.firewatchRequirement,
            form.areaInspected,
            form.fireExtinguisher,
            form.flammablesRemoved,
            form.sprinklerHeadsProtected,
            form.signature,
            sigImg
          ]}
        />;
      });
    case 'All Forms':
      return (
        <div className="all-forms">
          <HotworkForm />
          <PretaskForm />
        </div>
      );
    case 'Reports By Project':
      return (
        <div>
          <h2 className="data-viz-cards-header">Select a project</h2>
          <div className="data-viz-cards">
            {
              this.state.dataVizControl &&
                <div className="selected-div">
                  {
                    this.state.hotworkForms.concat(this.state.pretaskForms).map((form, index) => {
                      if (parseInt(this.state.dataVizControl.split(' - ')[1], 10) === form.project_id) {
                        const sigImg = new Image();
                        sigImg.src = form.signature;
                        return <Card
                          key={'form' + index}
                          header={
                            Object.keys(form).includes('firewatchRequirement') ? 'Hotwork Permit - ' + form.employee_name : 'Pretask Permit - ' + form.employee_name
                          }
                          cardClickHandler={this.mockCardClickHandler}
                          formDetails={[
                            form.employee_email,
                            form.project_id,
                            form.company,
                            form.date,
                            form.firewatchRequirement,
                            form.areaInspected,
                            form.fireExtinguisher,
                            form.flammablesRemoved,
                            form.sprinklerHeadsProtected,
                            form.signature,
                            sigImg
                          ]}
                        />;
                      }
                    })
                  }
                </div>
            }
            {
              this.aToZ(this.state.projects).map((project, index) => {
                return <Card
                  header={project.name}
                  body={<button onClick={() => this.reportsBySelectionHandler('Reports By Project - ' + project.id)}>Select</button>}
                  projectId={project.id}
                  key={'project' + index}
                  cardClickHandler={this.cardClickHandler}
                  reportsBySelectionHandler={this.reportsBySelectionHandler}
                />;
              })
            }
          </div>
        </div>
      );
    case 'Reports By Employee':
      return (
        <div>
          <h2 className="data-viz-cards-header">Select an employee</h2>
          <div className="data-viz-cards">
            {
              this.state.dataVizControl &&
                <div className="selected-div">
                  {
                    this.state.hotworkForms.concat(this.state.pretaskForms).map((form, index) => {
                      if (this.state.dataVizControl.split(' - ')[1] === form.employee_id) {
                        const sigImg = new Image();
                        sigImg.src = form.signature;
                        return <Card
                          key={'form' + index}
                          header={
                            Object.keys(form).includes('firewatchRequirement') ? 'Hotwork Permit - ' + form.employee_name : 'Pretask Permit - ' + form.employee_name
                          }
                          cardClickHandler={this.mockCardClickHandler}
                          formDetails={[
                            form.employee_email,
                            form.project_id,
                            form.company,
                            form.date,
                            form.firewatchRequirement,
                            form.areaInspected,
                            form.fireExtinguisher,
                            form.flammablesRemoved,
                            form.sprinklerHeadsProtected,
                            form.signature,
                            sigImg
                          ]}
                        />;
                      }
                    })
                  }
                </div>
            }
            {
              this.aToZ(this.state.employees).map((employee, index) => {
                return <Card
                  header={employee.name}
                  body={<button onClick={() => this.reportsBySelectionHandler('Reports By Employee - ' + employee.id)}>Select</button>}
                  employeeId={employee.id}
                  key={'employee' + index}
                  reportsBySelectionHandler={this.reportsBySelectionHandler}
                  cardClickHandler={this.cardClickHandler}
                />;
              })
            }
          </div>
        </div>
      );
    }
  }

  render() {
    return (
      <div className="data-viz">
        { this.renderWelcome(this.props.selectedControl) }
        { this.renderSelected(this.state.selectedInfo) }
        <div className="data-viz-cards">
          { this.cardSwitch(this.props.selectedControl) }
        </div>
      </div>

    );
  }
}

export default DataViz;
