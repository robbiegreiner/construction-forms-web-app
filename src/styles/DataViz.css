.data-viz {
  width: 50%;
  background-image: linear-gradient(to top left, #9419FF 0%, #33146F 100%);
  /* background-color: #9419FF; */
  display: flex;
  flex-direction: column;
  align-items: center;
}

.data-viz h1 {
  margin-top: 20px;
}

.data-viz-cards {
  display: flex;
  justify-content: space-around;
  flex-wrap: wrap;
  overflow: scroll;
}

.card {
  width: 32%;
  box-sizing: border-box;
  padding: 5px;
  border: 1px solid pink;
}
