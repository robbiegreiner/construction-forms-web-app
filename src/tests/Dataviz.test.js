/* eslint-disable no-unused-vars, no-undef, import/no-unresolved */
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import DataViz from '../components/DataViz';

configure({ adapter: new Adapter(), disableLifecycleMethods: true });

describe('DataViz tests', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<DataViz />);
  });

  it('should render without crashing', () => {
    shallow(<DataViz />);
  });

  it('should render with empty state', () => {
    const emptyState = {
      projects: [],
      employees: [],
      hotworkForms: [],
      pretaskForms: [],
      dataVizControl: null,
      selectedInfo: {}
    };

    expect(wrapper.state()).toEqual(emptyState);
  });

  it('should set selected view in state', () => {
    wrapper.instance().formsBySelector('Reports By Project - 10');

    expect(wrapper.state().dataVizControl).toEqual('Reports By Project - 10');

  });

  it('should set selected card in state', () => {
    const mockCardProps = {
      body: {},
      cardClickHandler: jest.fn(),
      formsBySelector: true,
      header: 'Confluence Condominiums',
      projectId: 10
    };

    wrapper.instance().cardClickHandler(mockCardProps);

    expect(wrapper.state().selectedInfo).toBe(mockCardProps);
  });

  it('should do nothing when clicked on a form by selector', () => {
    wrapper.instance().setState({ selectedInfo: 'Sirius' });
    wrapper.instance().mockCardClickHandler({ selectedInfo: 'Eye In The Sky' });
    expect(wrapper.state().selectedInfo).toEqual('Sirius');
  });

  it('should clear state when closeSelected', () => {
    wrapper.instance().setState({ selectedInfo: 'Sirius' });
    wrapper.instance().closeSelected();

    expect(wrapper.state().selectedInfo).toEqual({});
  });

  it('should sort an array of objects by name', () => {
    const arrayOfObjs = [
      { name: 'Jorge' },
      { name: 'Alan' },
      { name: 'Robbie' }
    ];

    const sortedArray = [
      { name: 'Alan' },
      { name: 'Jorge' },
      { name: 'Robbie' }
    ];

    expect(wrapper.instance().aToZ(arrayOfObjs)).toEqual(sortedArray);
  });

  it('should render a welcome page when no control is selected', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot when selectedControl is Projects', () => {
    const wrapper = shallow(<DataViz selectedControl='Projects' />);
    wrapper.instance().setState({
      projects: [
        { id: 1, location: 'Omaha', name: 'FNBO Building', public: true, union: true },
        { id: 2, location: 'Boston', name: 'Stadium', public: true, union: false }
      ]
    });
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot when selectedControl is Employees', () => {
    const wrapper = shallow(<DataViz selectedControl='Employees' />);
    wrapper.instance().setState({
      employees: [
        { id: 1, email: 'robbie@email.com', name: 'Robbie Griener', phone: '555-555-5556', position: 'The Boss' },
        { id: 2, email: 'mickjeets@email.com', name: 'Mick Jeets', phone: '444-444-4445', position: 'Underling' }
      ]
    });
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot when selectedControl is All Reports', () => {
    const wrapper = shallow(<DataViz selectedControl='All Reports' />);
    wrapper.instance().setState({
      hotworkForms: [
        {
          id: 1,
          areaInspected: true,
          company: 'Mortenson',
          date: '2018-01-07T21:18:12.993Z',
          employee_email: 'robbie@g.com',
          employee_id: '31',
          employee_name: 'Robbie',
          fireExtinguisher: true,
          firewatchRequirement: '30 minutes',
          flammablesRemoved: true,
          project_id: 11,
          signature: 'xxx'
        }
      ],
      pretaskForms: [
        {
          id: 1,
          adequateLighting: true,
          buddyAssignment: false,
          company: 'Mortenson',
          crewSize: 12,
          date: '2018-01-07T21:18:12.993Z',
          employee_email: 'robbie@g.com',
          employee_id: '31',
          employee_name: 'Robbie',
          equipmentShutdown: true,
          fluidDischarge: false,
          hazards: 'Overhead',
          impactOwner: false,
          lifting: false,
          msdsReviewed: true,
          planReview: true,
          project_id: 11,
          requireTraining: true,
          safetyLocations: false,
          signature: 'xxx'
        }
      ]
    });
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot when Reports By Project is selectedControl', () => {
    const wrapper = shallow(<DataViz selectedControl='Reports By Project' />);
    wrapper.instance().setState({
      employees: [
        { id: 1, email: 'dani@g.com', name: 'Dani Griener', location: 'Denver', position: 'Welder' }
      ],
      hotworkForms: [
        {
          id: 1,
          areaInspected: true,
          company: 'Mortenson',
          date: '2018-01-07T21:18:12.993Z',
          employee_email: 'robbie@g.com',
          employee_id: '31',
          employee_name: 'Robbie',
          fireExtinguisher: true,
          firewatchRequirement: '30 minutes',
          flammablesRemoved: true,
          project_id: 11,
          signature: 'xxx'
        }
      ]
    });
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should match snapshot when Reports By Employee is selectedControl', () => {
    const wrapper = shallow(<DataViz selectedControl='Reports By Employee' />);
    wrapper.instance().setState({
      employees: [
        { id: 1, location: 'Omaha', name: 'FNBO Building', public: true, union: true }
      ],
      hotworkForms: [
        {
          id: 1,
          areaInspected: true,
          company: 'Mortenson',
          date: '2018-01-07T21:18:12.993Z',
          employee_email: 'robbie@g.com',
          employee_id: '31',
          employee_name: 'Robbie',
          fireExtinguisher: true,
          firewatchRequirement: '30 minutes',
          flammablesRemoved: true,
          project_id: 11,
          signature: 'xxx'
        }
      ]
    });
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});
