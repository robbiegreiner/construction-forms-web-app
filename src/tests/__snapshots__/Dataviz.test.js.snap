// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataViz tests should match snapshot when Reports By Employee is selectedControl 1`] = `
<div
  className="data-viz"
>
  <div
    className="selected-div"
  />
  <div
    className="data-viz-cards"
  >
    <div>
      <h2
        className="data-viz-cards-header"
      >
        Select an employee
      </h2>
      <div
        className="data-viz-cards"
      >
        <Card
          body={
            <button
              onClick={[Function]}
            >
              Select
            </button>
          }
          cardClickHandler={[Function]}
          employeeId={1}
          formsBySelector={true}
          header="FNBO Building"
          key="employee0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`DataViz tests should match snapshot when Reports By Project is selectedControl 1`] = `
<div
  className="data-viz"
>
  <div
    className="selected-div"
  />
  <div
    className="data-viz-cards"
  >
    <div>
      <h2
        className="data-viz-cards-header"
      >
        Select a project
      </h2>
      <div
        className="data-viz-cards"
      />
    </div>
  </div>
</div>
`;

exports[`DataViz tests should match snapshot when selectedControl is All Reports 1`] = `
<div
  className="data-viz"
>
  <div
    className="selected-div"
  />
  <div
    className="data-viz-cards"
  >
    <Card
      cardClickHandler={[Function]}
      formDetails={
        Array [
          "robbie@g.com",
          11,
          "Mortenson",
          "2018-01-07T21:18:12.993Z",
          "30 minutes",
          true,
          true,
          true,
          undefined,
          "xxx",
          <img
            src="xxx"
          />,
        ]
      }
      header="Hotwork Permit - Robbie"
      key="form0"
    />
    <Card
      cardClickHandler={[Function]}
      formDetails={
        Array [
          "robbie@g.com",
          11,
          "Mortenson",
          "2018-01-07T21:18:12.993Z",
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          "xxx",
          <img
            src="xxx"
          />,
        ]
      }
      header="Pretask Permit - Robbie"
      key="form1"
    />
  </div>
</div>
`;

exports[`DataViz tests should match snapshot when selectedControl is Employees 1`] = `
<div
  className="data-viz"
>
  <div
    className="selected-div"
  />
  <div
    className="data-viz-cards"
  >
    <Card
      body="Underling"
      cardClickHandler={[Function]}
      employeeInfo={
        Array [
          "444-444-4445",
          "mickjeets@email.com",
        ]
      }
      header="Mick Jeets"
      key="employee0"
    />
    <Card
      body="The Boss"
      cardClickHandler={[Function]}
      employeeInfo={
        Array [
          "555-555-5556",
          "robbie@email.com",
        ]
      }
      header="Robbie Griener"
      key="employee1"
    />
  </div>
</div>
`;

exports[`DataViz tests should match snapshot when selectedControl is Projects 1`] = `
<div
  className="data-viz"
>
  <div
    className="selected-div"
  />
  <div
    className="data-viz-cards"
  >
    <Card
      body="Omaha"
      cardClickHandler={[Function]}
      header="FNBO Building"
      key="project0"
      projectInfo={
        Array [
          true,
          true,
        ]
      }
    />
    <Card
      body="Boston"
      cardClickHandler={[Function]}
      header="Stadium"
      key="project1"
      projectInfo={
        Array [
          true,
          false,
        ]
      }
    />
  </div>
</div>
`;

exports[`DataViz tests should render a welcome page when no control is selected 1`] = `
<div
  className="data-viz"
>
  <div
    className="welcome"
  >
    <h1>
      Welcome to Construction Forms!
    </h1>
    <div
      className="logo"
    />
  </div>
  <div
    className="selected-div"
  />
  <div
    className="data-viz-cards"
  />
</div>
`;
